core/testcontainers/core/container.py

```diff
@@ -56,11 +63,14 @@ def maybe_emulate_amd64(self) -> 'DockerContainer':
         return self
 
     def start(self) -> 'DockerContainer':
+        if self.ryuk and not self.image == REAPER_IMAGE:
+            logger.debug("Creating Ryuk container")
+            Reaper.get_instance()
         logger.info("Pulling image %s", self.image)
         docker_client = self.get_docker_client()
         self._container = docker_client.run(
             self.image, command=self._command, detach=True, environment=self.env, ports=self.ports,
-            name=self._name, volumes=self.volumes, **self._kwargs
+            name=self._name, volumes=self.volumes, cleanup_on_exit=not self.ryuk, **self._kwargs
```

Comment:
I want to specifically disable the behaviour where the container gets killed when the memory is claimed by the GC (through the __del__ dunder). 

Killing the container on GC removes the possibility to reuse the container within 10 seconds, which is very useful when watching a project and running tests on file change or during development with Django / FastAPI which performs a lot of restarts. Having to wait for a container which maybe takes 2 - 10 seconds to spin up for every file change on hot reloading setups makes for a very bad dev experience. 

Also, most of the other implementations rely solely on Ryuk to cleanup after usage, so not killing the container on GC is closer to the behaviour of the other implementations

Reply:
Reading your other suggestion of making Ryuk enabled by default, what do you think about removing the GC during `__del__` completely?

Reply:
I see, I am fine with whatever works best for the Python community. I was just thinking how other implementations do it.

> Reading your other suggestion of making Ryuk enabled by default, what do you think about removing the GC during `__del__` completely?

I have no strong opinion here. Having the clean up immediately is sometimes nice to have, e.g. .NET does it too. (similar)

Reply:
Having given it a bit more thought, I think we can have a very nice API that both supports controlling the lifecycle of a container explicitly, and making sure that containers are always cleaned up by Ryuk:

We enable Ryuk by default. This makes sure that all containers are cleaned up, even if you forget to do it explicitly or implicitly (see below). At the same time, we remove the cleanup logic from `__del__` while still keeping it in the `__exit__` dunder, in order to still support the `with` usage (which I guess most people are using due to the examples in the README and other places). It also plays nice with the move FastAPI and other modern frameworks are currently make towards using `contextmanagers` for dependency management / injection.

Having `__del__` clean up everything is quite limiting in terms of usage, as one always has to keep a reference to the container in your application code if you want to keep it alive. Given the comments currently in the `__del__` method I think the goal was to perform cleanup no matter what, which we now will use Ryuk for instead.

The following use cases as thus supported:
```
with DockerContainer("postgres") as container:
    # some logic here

# end of context: container is killed by `__exit__` method
```

```
container = DockerContainer("postgres").start()
# some logic here
container.stop()
```

```
container = DockerContainer("postgres").start()
# some logic here

# You forget to .stop() the container, but Ryuk kills it for you 10 seconds after your process stopped.
```

Reply:
Sounds reasonable, but I cannot add much more because I am not very familiar with the overall Python implementation. There is another use case: Ryuk will do the clean up even if the test process crashes, etc.